twitter_pr.rs

```diff
@@ -0,0 +1,96 @@
+extern crate reqwest;
+
+use std::sync::mpsc::{Receiver, channel};
+use serde::{Serialize, Deserialize};
+use tokio::stream::StreamExt;
+use tokio::io::{AsyncBufReadExt, StreamReader};
+use serde_json::{Deserializer, StreamDeserializer};
+use std::io::Write;
+
+pub struct TwitterApi {
+    pub bearer_api_token: String
+}
+
+struct RequestStream {
+    receiver: Receiver<bytes::Bytes>
+}
+
+impl std::io::Read for RequestStream {
+    fn read(&mut self, mut buf: &mut [u8]) -> std::io::Result<usize> {
+        match self.receiver.recv() {
+            Ok(next) => {
+                let vec = next.to_vec();
+                let vec_len = vec.len();
+                println!("Incoming -> {}", String::from_utf8(vec.clone()).unwrap_or("err".to_string()));
+                buf.write_all(vec.as_slice());
+                return Ok(vec_len);
+            }
+            Err(_e) => panic!("error!")
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Tweet {
+    pub text: String,
+}
+
+#[derive(Deserialize)]
+struct Wrapper<T> {
+    data: T,
+}
+
+impl TwitterApi {
+    pub fn new(api_token: String) -> TwitterApi {
+        TwitterApi {
+            bearer_api_token: api_token
+        }
+    }
+
+    pub async fn listen_for_tweets(&self) -> Receiver<Tweet> {
+        let (sender, receiver) = channel::<Tweet>();
+
+        let client = reqwest::Client::new();
+        let url = reqwest::Url::parse("https://api.twitter.com/2/tweets/search/stream").unwrap();
```

Comment:
[reqwest's get](https://docs.rs/reqwest/0.10.8/reqwest/fn.get.html) takes an `Into<Url>`, which means anything that can be turned into a URL by way of the `Into` trait. The documentation doesn't make this clear (a general gripe with the Into system), but it turns out an `&str` is "Into" Url: https://docs.rs/reqwest/0.10.8/src/reqwest/into_url.rs.html#8. So rather than parsing the URL explicitly, you can just pass the URL as a `&str` to get().